# A snakemake workflow to generate multiple synthetic datasets using aMGSIM

This workflow is designed to help in the generation of multiple synthetic datasets using [aMGSIM](https://github.com/genomewalker/aMGSIM). The workflow needs the results generated after processing a BAM file through [filterBAM](https://github.com/genomewalker/bam-filter) and [metaDMG](https://metadmg-dev.github.io/metaDMG-core/). An example of the config file is [here](config/config.yaml).

The [sample TSV](assets/metaDMGsamplelist-test.tsv) file must contain the following fields:
- **label**: The name of the sample
- **short_label**: A short name for the sample (if not provided will autogenerate a short name using the MD5 of the label)
- **libprep**: The type of library preparation to use (**single**-stranded or **double**-stranded)
- **read_length_freqs_file**: A JSON file with the empirical read length frequencies for the sample estimated by filterBAM
- **mapping_stats_filtered_file**: A TSV file with the filtered mapping statistics generated by filterBAM
- **metadmg_results**: The TSV generated by metaDMG when analyzing the filtered BAM file using the metaDMG's local mode
- **metadmg_misincorporations**: The misincorporations generated by metaDMG when analyzing the filtered BAM file using the metaDMG's local mode

Once the workflow is done will produce the following folders:

```
Cave-22/single/single/1000000
├── Cave-22.communities.tsv
├── Cave-22.filepaths.tsv
├── Cave-22.genome-compositions.tsv
├── Cave-22.json
├── Cave-22.tsv
├── Cave-22_read-files.json
├── ag-config.yaml
├── ar-config.yaml
├── fb-config.yaml
└── reads
    ├── Cave-22_art.fq.gz
    ├── Cave-22_deamSim.fa.gz
    └── Cave-22_fragSim.fa.gz
```